Sources/ParseSwift/Coding/ParseEncoder.swift

```diff
@@ -0,0 +1,259 @@
+//
+//  ParseEncoder.swift
+//  ParseSwift
+//
+//  Created by Pranjal Satija on 7/20/20.
+//  Copyright © 2020 Parse. All rights reserved.
+//
+
+// This rule doesn't allow types with underscores in their names.
+// swiftlint:disable type_name
+
+import Foundation
+
+// MARK: ParseEncoder
+public struct ParseEncoder {
+    let dateEncodingStrategy: AnyCodable.DateEncodingStrategy?
+    let jsonEncoder: JSONEncoder
+    let skippedKeys: Set<String>
+
+    init(
+        dateEncodingStrategy: AnyCodable.DateEncodingStrategy? = nil,
+        jsonEncoder: JSONEncoder = JSONEncoder(),
+        skippingKeys: Set<String> = []
+    ) {
+        self.dateEncodingStrategy = dateEncodingStrategy
+        self.jsonEncoder = jsonEncoder
+        self.skippedKeys = skippingKeys
+    }
+
+    func encodeToDictionary<T: Encodable>(_ value: T) throws -> [AnyHashable: Any] {
+        let encoder = _ParseEncoder(codingPath: [], dictionary: NSMutableDictionary(), skippingKeys: skippedKeys)
+        try value.encode(to: encoder)
+
+        // swiftlint:disable:next force_cast
+        return encoder.dictionary as! [AnyHashable: Any]
+    }
+
+    func encode<T: Encodable>(_ value: T) throws -> Data {
+        let dictionary = try encodeToDictionary(value)
+        return try jsonEncoder.encode(AnyCodable(dictionary, dateEncodingStrategy: dateEncodingStrategy!))
+    }
+
+    func encode<T: Encodable>(_ array: [T]) throws -> Data {
+        let dictionaries = try array.map { try encodeToDictionary($0) }
+        return try jsonEncoder.encode(AnyCodable(dictionaries, dateEncodingStrategy: dateEncodingStrategy!))
+    }
+}
+
+// MARK: _ParseEncoder
+internal struct _ParseEncoder: Encoder {
+    let codingPath: [CodingKey]
+    let dictionary: NSMutableDictionary
+    let skippedKeys: Set<String>
+    let userInfo: [CodingUserInfoKey: Any] = [:]
+
+    init(codingPath: [CodingKey], dictionary: NSMutableDictionary, skippingKeys: Set<String>) {
```

Comment:
So `NSMutableDictionary` and `Dictionary` actually have very different behavior; the `NS` version is a reference type, and it's still around for cases where you actually want reference types. In this case, we need it to work as a reference type because the dictionary gets passed around to a few different places as encoding occurs, and we want everyone writing into the same dictionary.